libs/vkd3d/opacity_micromap.c

```diff
@@ -0,0 +1,286 @@
+
+        RT_TRACE("  Count: %u\n", histogram_entry->Count);
+        RT_TRACE("  Subdivision level: %u\n", histogram_entry->SubdivisionLevel);
+        RT_TRACE("  Format: %s\n", debug_omm_format(histogram_entry->Format));
+    }
+
+    build_info->pUsageCounts = usages;
+    build_info->data.deviceAddress = desc->InputBuffer;
+    build_info->triangleArray.deviceAddress = desc->PerOmmDescs.StartAddress;
+    build_info->triangleArrayStride = desc->PerOmmDescs.StrideInBytes;
+
+    RT_TRACE(" IBO VA: %"PRIx64"\n", desc->InputBuffer);
+    RT_TRACE(" Triangles VA: %"PRIx64"\n", desc->PerOmmDescs.StartAddress);
+    RT_TRACE(" Triangles stride: %"PRIu64" bytes\n", desc->PerOmmDescs.StrideInBytes);
+
+    RT_TRACE("=====================\n");
+    return true;
+}
+
+static void vkd3d_opacity_micromap_end_barrier(struct d3d12_command_list *list)
+{
+    /* We resolve the query in TRANSFER, but DXR expects UNORDERED_ACCESS. */
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    VkDependencyInfo dep_info;
+    VkMemoryBarrier2 barrier;
+
+    memset(&barrier, 0, sizeof(barrier));
+    barrier.sType = VK_STRUCTURE_TYPE_MEMORY_BARRIER_2;
+    barrier.srcStageMask = VK_PIPELINE_STAGE_2_COPY_BIT;
+    barrier.srcAccessMask = VK_ACCESS_2_TRANSFER_WRITE_BIT;
+    barrier.dstStageMask = VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT;
+
+    memset(&dep_info, 0, sizeof(dep_info));
+    dep_info.sType = VK_STRUCTURE_TYPE_DEPENDENCY_INFO;
+    dep_info.memoryBarrierCount = 1;
+    dep_info.pMemoryBarriers = &barrier;
+
+    VK_CALL(vkCmdPipelineBarrier2(list->cmd.vk_command_buffer, &dep_info));
+}
+
+void vkd3d_opacity_micromap_write_postbuild_info(
+        struct d3d12_command_list *list,
+        const D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_DESC *desc,
+        VkDeviceSize desc_offset,
+        VkMicromapEXT vk_opacity_micromap)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    const struct vkd3d_unique_resource *resource;
+    VkQueryPool vk_query_pool;
+    VkQueryType vk_query_type;
+    uint32_t vk_query_index;
+    VkDeviceSize stride;
+    uint32_t type_index;
+    VkBuffer vk_buffer;
+    uint32_t offset;
+
+    resource = vkd3d_va_map_deref(&list->device->memory_allocator.va_map, desc->DestBuffer);
+    if (!resource)
+    {
+        ERR("Invalid resource.\n");
+        return;
+    }
+
+    vk_buffer = resource->vk_buffer;
+    offset = desc->DestBuffer - resource->va;
+    offset += desc_offset;
+
+    switch (desc->InfoType)
+    {
+        case D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_COMPACTED_SIZE:
+            vk_query_type = VK_QUERY_TYPE_MICROMAP_COMPACTED_SIZE_EXT;
+            type_index = VKD3D_QUERY_TYPE_INDEX_OMM_COMPACTED_SIZE;
+            stride = sizeof(uint64_t);
+            break;
+        case D3D12_RAYTRACING_ACCELERATION_STRUCTURE_POSTBUILD_INFO_SERIALIZATION:
+            vk_query_type = VK_QUERY_TYPE_MICROMAP_SERIALIZATION_SIZE_EXT;
```

Comment:
Just flag this FIXME and write 0 too. We know it's broken on NV atm, and it's not a very exciting query anyway.